File: sdks/python/hatchet_sdk/features/metrics.py
import asyncio
from datetime import datetime, timedelta, timezone
from typing import Any

from pydantic import BaseModel

from hatchet_sdk.clients.rest.api.task_api import TaskApi
from hatchet_sdk.clients.rest.api.tenant_api import TenantApi
from hatchet_sdk.clients.rest.api_client import ApiClient
from hatchet_sdk.clients.rest.models.task_stat import TaskStat
from hatchet_sdk.clients.rest.tenacity_utils import tenacity_retry
from hatchet_sdk.clients.v1.api_client import BaseRestClient


class TaskMetrics(BaseModel):
    cancelled: int
    completed: int
    failed: int
    queued: int
    running: int


class MetricsClient(BaseRestClient):
    """
    The metrics client is a client for reading metrics out of Hatchet's metrics API.
    """

    def _taskapi(self, client: ApiClient) -> TaskApi:
        return TaskApi(client)

    def _ta(self, client: ApiClient) -> TenantApi:
        return TenantApi(client)

    def get_queue_metrics(
        self,
    ) -> dict[str, Any]:
        """
        Retrieve the current queue metrics for the tenant.

        :return: The current queue metrics
        """
        with self.client() as client:
            tenant_get_step_run_queue_metrics = tenacity_retry(
                self._ta(client).tenant_get_step_run_queue_metrics,
                self.client_config.tenacity,
            )
            return (
                tenant_get_step_run_queue_metrics(
                    tenant=self.client_config.tenant_id,
                ).queues
            ) or {}

    async def aio_get_queue_metrics(
        self,
    ) -> dict[str, Any]:
        """
        Retrieve the current queue metrics for the tenant.

        :return: The current queue metrics
        """

        return await asyncio.to_thread(self.get_queue_metrics)

    def scrape_tenant_prometheus_metrics(
        self,
    ) -> str:
        """
        Scrape Prometheus metrics for the tenant. Returns the metrics in Prometheus text format.

        :return: The metrics, returned in Prometheus text format
        """
        with self.client() as client:
            tenant_get_prometheus_metrics = tenacity_retry(
                self._ta(client).tenant_get_prometheus_metrics,
                self.client_config.tenacity,
            )
            return tenant_get_prometheus_metrics(
                tenant=self.client_config.tenant_id,
            )

    async def aio_scrape_tenant_prometheus_metrics(
        self,
    ) -> str:
        """
        Scrape Prometheus metrics for the tenant. Returns the metrics in Prometheus text format.

        :return: The metrics, returned in Prometheus text format
        """

        return await asyncio.to_thread(self.scrape_tenant_prometheus_metrics)

    def get_task_stats(self) -> dict[str, TaskStat]:
        with self.client() as client:
            get_task_stats = tenacity_retry(
                self._ta(client).tenant_get_task_stats, self.client_config.tenacity
            )
            return get_task_stats(
                tenant=self.client_config.tenant_id,
            )

    async def aio_get_task_stats(self) -> dict[str, TaskStat]:
        return await asyncio.to_thread(self.get_task_stats)

    def get_task_metrics(
        self,
        since: datetime | None = None,
        until: datetime | None = None,
        workflow_ids: list[str] | None = None,
        parent_task_external_id: str | None = None,
        triggering_event_external_id: str | None = None,
    ) -> TaskMetrics:
        """
        Retrieve task metrics, grouped by status (queued, running, completed, failed, cancelled).

        :param since: Start time for the metrics query (defaults to the past day if unset)
        :param until: End time for the metrics query
        :param workflow_ids: List of workflow IDs to filter the metrics by
        :param parent_task_external_id: ID of the parent task to filter by (note that parent task here refers to the task that spawned this task as a child)
        :param triggering_event_external_id: ID of the triggering event to filter by
        :return: Task metrics
        """

        since = since or datetime.now(timezone.utc) - timedelta(days=1)
        until = until or datetime.now(timezone.utc)
        with self.client() as client:
            v1_task_list_status_metrics = tenacity_retry(
                self._taskapi(client).v1_task_list_status_metrics,
                self.client_config.tenacity,
            )
            metrics = {
                m.status.name.lower(): m.count
                for m in v1_task_list_status_metrics(
                    tenant=self.client_config.tenant_id,
                    since=since,
                    until=until,
                    workflow_ids=workflow_ids,
                    parent_task_external_id=parent_task_external_id,
                    triggering_event_external_id=triggering_event_external_id,
                )
            }

            return TaskMetrics.model_validate(metrics)

    async def aio_get_task_metrics(
        self,
        since: datetime | None = None,
        until: datetime | None = None,
        workflow_ids: list[str] | None = None,
        parent_task_external_id: str | None = None,
        triggering_event_external_id: str | None = None,
    ) -> TaskMetrics:
        """
        Retrieve task metrics, grouped by status (queued, running, completed, failed, cancelled).

        :param since: Start time for the metrics query (defaults to the past day if unset)
        :param until: End time for the metrics query
        :param workflow_ids: List of workflow IDs to filter the metrics by
        :param parent_task_external_id: ID of the parent task to filter by (note that parent task here refers to the task that spawned this task as a child)
        :param triggering_event_external_id: ID of the triggering event to filter by
        :return: Task metrics
        """
        return await asyncio.to_thread(
            self.get_task_metrics,
            since,
            until,
            workflow_ids,
            parent_task_external_id,
            triggering_event_external_id,
        )

File: sdks/ruby/src/lib/hatchet/features/metrics.rb
# frozen_string_literal: true

require "time"

module Hatchet
  module Features
    # Simple struct for task metrics grouped by status
    TaskMetrics = Struct.new(:cancelled, :completed, :failed, :queued, :running, keyword_init: true)

    # Metrics client for reading metrics out of Hatchet's metrics API
    #
    # This class provides a high-level interface for retrieving queue metrics,
    # Prometheus metrics, task statistics, and task metrics from the Hatchet system.
    #
    # @example Getting queue metrics
    #   metrics = metrics_client.get_queue_metrics
    #
    # @example Getting task metrics
    #   task_metrics = metrics_client.get_task_metrics(
    #     since: Time.now - 86400,
    #     workflow_ids: ["wf-1"]
    #   )
    #
    # @since 0.1.0
    class Metrics
      # Initializes a new Metrics client instance
      #
      # @param rest_client [Object] The configured REST client for API communication
      # @param config [Hatchet::Config] The Hatchet configuration containing tenant_id and other settings
      # @return [void]
      # @since 0.1.0
      def initialize(rest_client, config)
        @rest_client = rest_client
        @config = config
        @task_api = HatchetSdkRest::TaskApi.new(rest_client)
        @tenant_api = HatchetSdkRest::TenantApi.new(rest_client)
      end

      # Retrieve the current queue metrics for the tenant
      #
      # @return [Hash] The current queue metrics
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   queues = metrics_client.get_queue_metrics
      def get_queue_metrics
        result = @tenant_api.tenant_get_step_run_queue_metrics(@config.tenant_id)
        result.queues || {}
      end

      # Scrape Prometheus metrics for the tenant
      #
      # @return [String] The metrics in Prometheus text format
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   prometheus_text = metrics_client.scrape_tenant_prometheus_metrics
      def scrape_tenant_prometheus_metrics
        @tenant_api.tenant_get_prometheus_metrics(@config.tenant_id)
      end

      # Get task statistics for the tenant
      #
      # @return [Object] The task statistics
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   stats = metrics_client.get_task_stats
      def get_task_stats
        @tenant_api.tenant_get_task_stats(@config.tenant_id)
      end

      # Retrieve task metrics grouped by status (queued, running, completed, failed, cancelled)
      #
      # @param since [Time, nil] Start time for the metrics query (defaults to the past day if unset)
      # @param until_time [Time, nil] End time for the metrics query
      # @param workflow_ids [Array<String>, nil] List of workflow IDs to filter the metrics by
      # @param parent_task_external_id [String, nil] ID of the parent task to filter by
      # @param triggering_event_external_id [String, nil] ID of the triggering event to filter by
      # @return [TaskMetrics] Task metrics with counts per status
      # @raise [HatchetSdkRest::ApiError] If the API request fails
      # @example
      #   metrics = metrics_client.get_task_metrics(
      #     since: Time.now - 86400,
      #     workflow_ids: ["wf-1"]
      #   )
      #   puts "Completed: #{metrics.completed}, Failed: #{metrics.failed}"
      def get_task_metrics(since: nil, until_time: nil, workflow_ids: nil, parent_task_external_id: nil, triggering_event_external_id: nil)
        since_time = since || (Time.now - (24 * 60 * 60))
        until_val = until_time || Time.now

        result = @task_api.v1_task_list_status_metrics(
          @config.tenant_id,
          since_time.utc.iso8601,
          {
            _until: until_val.utc.iso8601,
            workflow_ids: workflow_ids,
            parent_task_external_id: parent_task_external_id,
            triggering_event_external_id: triggering_event_external_id,
          },
        )

        # Build metrics hash from the status metric objects
        metrics = { cancelled: 0, completed: 0, failed: 0, queued: 0, running: 0 }

        if result.is_a?(Array)
          result.each do |m|
            status_name = m.respond_to?(:status) ? m.status.to_s.downcase.to_sym : nil
            count = m.respond_to?(:count) ? m.count : 0
            metrics[status_name] = count if metrics.key?(status_name)
          end
        end

        TaskMetrics.new(**metrics)
      end
    end
  end
end
